### ISSUE TITLE
Question on cross_validation and changepoints

### ISSUE THREAD

**Issue by SoccerGeekPhD** [facebook/prophet#1529]
I am using Google Mobility data to create models of the current impact of social distancing.  I also use the state policy data at https://github.com/COVID19StatePolicy/SocialDistancing for a list of Changepoints .

The Changepoints include relaxation/expiration of orders some of which are beyond the data collection period.  I remove these late dates from the Changepoints and successfully fit a model.

The issue arises when I try to call cross_validation on the model.  I wish to use the initial period as all the data except the last 28 days
```
for( cp in scales){
    pr <- prophet(df =curr.dt[, .(ds,y)], holidays=hDays, 
                  changepoints = .changeDates,
                  n.changepoints = .n.changePoints, # only used when .changeDates=NULL
                  changepoint.prior.scale = cp,
                  daily.seasonality = FALSE, 
                  weekly.seasonality = TRUE,
                  yearly.seasonality = FALSE
    )
    
    p.cv <- cross_validation(pr, initial = curr.dt[ds <= lastDay-days(28),.N], 
                             period = 2,
                             horizon = 21, units = 'days')
# more code to chose best cp value using MAPE
}
```
when I run this I get an error "Changepoints must fall within training data."  This comes from cross_validation, not prophet call.

As the initial model, pr,  fit correctly, then does this mean that the Changepoints must fall in the initial window?  Is that what is happening here?

Is the correct thing to do to restrict the Changepoints to pick the parameters, then refit one more time after including the last Changepoints?  

To me, cross_validation should warn the user and trim Changepoints to the initial period.

**Comment by bletham** [comment]
For each cutoff of cross validation, a new model is constructed and when this is done the changepoints actually are trimmed to the initial period. That happens here:
https://github.com/facebook/prophet/blob/9a27386cccc01f1b7529bd613da64562bef92f34/R/R/diagnostics.R#L180-L185

So what you're trying to do should work and I'm not sure why you're running into an error. Can you post the full traceback, and see if there is a cleaner example you can produce that recreates the issues? When I try this with data from the quickstart, it works:
```R
library(prophet)
df <- read.csv('../examples/example_wp_log_peyton_manning.csv')
changepoints <- as.Date(c(
    '2008-01-01', '2009-01-01', '2010-01-01', '2011-01-01', '2012-01-01', '2013-01-01',
    '2014-01-01', '2015-01-01', '2016-01-01'
))
m <- prophet(df, changepoints = changepoints)

df.cv <- cross_validation(m, initial = 1500, period = 365, horizon = 365, units = 'days')
```
It prints out `Making 4 forecasts with cutoffs between 2012-01-21 and 2015-01-20`, so there are changepoints after the initial cutoff that are being trimmed as expected.

**Comment by SoccerGeekPhD** [comment]
Here's the example.  Two other thoughts come to mind, using Date() class or using data.tables.
```
require(data.table)
require(prophet)

curr.dt <- structure(list(ds = structure(c(18309, 18310, 18311, 18312, 18317, 
                                           18319, 18324, 18326, 18331, 18333, 18337, 18338, 18339, 18340, 
                                           18341, 18344, 18345, 18346, 18347, 18348, 18351, 18352, 18353, 
                                           18354, 18355, 18358, 18359, 18360, 18361, 18362, 18365, 18366, 
                                           18367, 18368, 18369, 18372, 18373, 18374, 18375, 18376, 18379, 
                                           18380, 18381, 18382, 18383, 18386, 18387, 18388, 18389, 18390, 
                                           18393, 18394, 18395, 18396, 18397, 18400, 18401, 18402, 18403, 
                                           18404), class = "Date"), 
                                        y = c(-20, 0, -5, 0, -2, -4, -2, -2, 
                                        -2, 0, -5, -11, -7, -13, -29, -20, -17, -20, -22, -26, -34, -35, 
                                        -30, -27, -38, -30, -30, -27, -33, -39.64, -37, -34, -25, -29, 
                                        -32, -30, -39.64, -33, -32, -31, -30, -32, -25, -28, -38, -25, 
                                        -28, -22, -26, -26, -25, -36, -29, -28, -27, -22, -27, -28, -31, 
                                        -33)), row.names = c(NA, -60L), class = c("data.table", "data.frame")
                     )

ctyChangeDates<- structure(c(18334, 18338, 18339, 18349, 18400), class = "Date")

hDays <- structure(list(holiday = c("PresidentsDay", "StPatricksDay", 
                                    "PalmSunday", "GoodFriday", "Easter", 
                                    "MemorialDay"), 
                            ds = structure(
                                      c(18309,18338, 18357, 18364, 18364, 18407), 
                                    class = "Date")), row.names = c(NA,-6L), 
                   class = c("data.table", "data.frame"))

lastDay <- curr.dt[, max(ds)]

pr <- prophet(df =curr.dt[, .(ds,y)], holidays=hDays, 
              changepoints = ctyChangeDates,
              #n.changepoints = .n.changePoints, # only used when .changeDates=NULL
              changepoint.prior.scale = 0.2,
              daily.seasonality = FALSE, 
              weekly.seasonality = TRUE,
              yearly.seasonality = FALSE
)

p.cv <- cross_validation(pr, initial = curr.dt[ds <= lastDay-28,.N], 
                         period = 2,
                         horizon = 21, units = 'days')
```

And the traceback
```r
5: stop("Changepoints must fall within training data.")
4: set_changepoints(m)
3: (function (m, df, ...) 
   {
       if (!is.null(m$history)) {
           stop("Prophet object can only be fit once. Instantiate a new object.")
       }
       if (!(exists("ds", where = df)) | !(exists("y", where = df))) {
           stop(paste("Dataframe must have columns 'ds' and 'y' with the dates and values", 
               "respectively."))
       }
       history <- df %>% dplyr::filter(!is.na(y))
       if (nrow(history) < 2) {
           stop("Dataframe has less than 2 non-NA rows.")
       }
       m$history.dates <- sort(set_date(df$ds))
       out <- setup_dataframe(m, history, initialize_scales = TRUE)
       history <- out$df
       m <- out$m
       m$history <- history
       m <- set_auto_seasonalities(m)
       out2 <- make_all_seasonality_features(m, history)
       m <- out2$m
       seasonal.features <- out2$seasonal.features
       prior.scales <- out2$prior.scales
       component.cols <- out2$component.cols
       m$train.component.cols <- component.cols
       m$component.modes <- out2$modes
       m$fit.kwargs <- list(...)
       m <- set_changepoints(m)
       dat <- list(T = nrow(history), K = ncol(seasonal.features), 
           S = length(m$changepoints.t), y = history$y_scaled, t = history$t, 
           t_change = array(m$changepoints.t), X = as.matrix(seasonal.features), 
           sigmas = array(prior.scales), tau = m$changepoint.prior.scale, 
           trend_indicator = as.numeric(m$growth == "logistic"), 
           s_a = array(component.cols$additive_terms), s_m = array(component.cols$multiplicative_terms))
       if (m$growth == "linear") {
           dat$cap <- rep(0, nrow(history))
           kinit <- linear_growth_init(history)
       }
       else {
           dat$cap <- history$cap_scaled
           kinit <- logistic_growth_init(history)
       }
       if (exists(".prophet.stan.model", where = prophet_model_env)) {
           model <- get(".prophet.stan.model", envir = prophet_model_env)
       }
       else {
           model <- stanmodels$prophet
       }
       stan_init <- function() {
           list(k = kinit[1], m = kinit[2], delta = array(rep(0, 
               length(m$changepoints.t))), beta = array(rep(0, ncol(seasonal.features))), 
               sigma_obs = 1)
       }
       if (min(history$y) == max(history$y)) {
           m$params <- stan_init()
           m$params$sigma_obs <- 0
           n.iteration <- 1
       }
       else if (m$mcmc.samples > 0) {
           args <- list(object = model, data = dat, init = stan_init, 
               iter = m$mcmc.samples)
           args <- utils::modifyList(args, list(...))
           m$stan.fit <- do.call(rstan::sampling, args)
           m$params <- rstan::extract(m$stan.fit)
           n.iteration <- length(m$params$k)
       }
       else {
           args <- list(object = model, data = dat, init = stan_init, 
               algorithm = if (dat$T < 100) {
                   "Newton"
               } else {
                   "LBFGS"
               }, iter = 10000, as_vector = FALSE)
           args <- utils::modifyList(args, list(...))
           m$stan.fit <- do.call(rstan::optimizing, args)
           if (m$stan.fit$return_code != 0) {
               message("Optimization terminated abnormally. Falling back to Newton optimizer.")
               args$algorithm = "Newton"
               m$stan.fit <- do.call(rstan::optimizing, args)
           }
           m$params <- m$stan.fit$par
           n.iteration <- 1
       }
       for (name in c("delta", "beta")) {
           m$params[[name]] <- matrix(m$params[[name]], nrow = n.iteration)
       }
       for (name in c("k", "m", "sigma_obs")) {
           m$params[[name]] <- c(m$params[[name]])
       }
       if (m$n.changepoints == 0) {
           m$params$k <- m$params$k + m$params$delta[, 1]
           m$params$delta <- matrix(rep(0, length(m$params$delta)), 
               nrow = n.iteration)
       }
       return(m)
   })(m = list(growth = "linear", changepoints = c(1584057600, 1584403200, 
   1584489600, 1585353600), n.changepoints = 4L, changepoint.range = 0.8, 
       yearly.seasonality = FALSE, weekly.seasonality = FALSE, daily.seasonality = FALSE, 
       holidays = list(holiday = c("PresidentsDay", "StPatricksDay", 
       "PalmSunday", "GoodFriday", "Easter", "MemorialDay"), ds = c(18309, 
       18338, 18357, 18364, 18364, 18407)), seasonality.mode = "additive", 
       seasonality.prior.scale = 10, changepoint.prior.scale = 0.2, 
       holidays.prior.scale = 10, mcmc.samples = 0, interval.width = 0.8, 
       uncertainty.samples = 1000, specified.changepoints = TRUE, 
       start = NULL, y.scale = NULL, logistic.floor = FALSE, t.scale = NULL, 
       changepoints.t = NULL, seasonalities = list(weekly = list(
           period = 7, fourier.order = 3, prior.scale = 10, mode = "additive", 
           condition.name = NULL)), extra_regressors = list(), stan.fit = NULL, 
       params = list(), history = NULL, history.dates = NULL, train.holiday.names = NULL, 
       train.component.cols = NULL, component.modes = NULL, fit.kwargs = list()), 
       df = list(ds = c(1581897600, 1581984000, 1582070400, 1582156800, 
       1582588800, 1582761600, 1583193600, 1583366400, 1583798400, 
       1583971200, 1584316800, 1584403200, 1584489600, 1584576000, 
       1584662400, 1584921600, 1585008000, 1585094400, 1585180800, 
       1585267200), y = c(-20, 0, -5, 0, -2, -4, -2, -2, -2, 0, 
       -5, -11, -7, -13, -29, -20, -17, -20, -22, -26), floor = c(0, 
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), 
           t = c(0, 0.0105263157894737, 0.0210526315789474, 0.0315789473684211, 
           0.0842105263157895, 0.105263157894737, 0.157894736842105, 
           0.178947368421053, 0.231578947368421, 0.252631578947368, 
           0.294736842105263, 0.305263157894737, 0.315789473684211, 
           0.326315789473684, 0.336842105263158, 0.368421052631579, 
           0.378947368421053, 0.389473684210526, 0.4, 0.410526315789474
           ), y_scaled = c(-0.504540867810293, 0, -0.126135216952573, 
           0, -0.0504540867810293, -0.100908173562059, -0.0504540867810293, 
           -0.0504540867810293, -0.0504540867810293, 0, -0.126135216952573, 
           -0.277497477295661, -0.176589303733602, -0.32795156407669, 
           -0.731584258324924, -0.504540867810293, -0.428859737638749, 
           -0.504540867810293, -0.554994954591322, -0.65590312815338
           )))
2: do.call(fit.prophet, fit.args)
1: cross_validation(pr, initial = curr.dt[ds <= lastDay - 28, .N], 
       period = 2, horizon = 21, units = "days")
```

**Comment by bletham** [comment]
Thanks, I see what's happening here now.
As you can see in the printout, the first generated cutoff is 2020-03-28. When the corresponding CV model is created, the data is trimmed to points <= cutoff:
https://github.com/facebook/prophet/blob/20d6afba8b354176d323701e0e6a3137796236d5/R/R/diagnostics.R#L130

and the changepoints are trimmed to points <= cutoff:
https://github.com/facebook/prophet/blob/20d6afba8b354176d323701e0e6a3137796236d5/R/R/diagnostics.R#L184

The full model has changepoints
```
[1] "2020-03-13 GMT" "2020-03-17 GMT" "2020-03-18 GMT" "2020-03-28 GMT"
[5] "2020-05-18 GMT"
```
so the first CV model (with cutoff 2020-03-28) has changepoints
```
[1] "2020-03-13 GMT" "2020-03-17 GMT" "2020-03-18 GMT" "2020-03-28 GMT"
```
In the dataset, however, there is a gap between 2020-03-27 and 2020-03-30:
```
19 2020-03-26 -22.00     0 0.40000000 -0.55499495
20 2020-03-27 -26.00     0 0.41052632 -0.65590313
21 2020-03-30 -34.00     0 0.44210526 -0.85771948
22 2020-03-31 -35.00     0 0.45263158 -0.88294652
```
So the history for the dataset ends at 2020-03-27. Thus, there is a changepoint at 2020-03-28 but the latest point in the data is 2020-03-27.

This is a bug, it's something that should work. The fix will be here the changepoints should be trimmed according to the data and not just the cutoff:
https://github.com/facebook/prophet/blob/20d6afba8b354176d323701e0e6a3137796236d5/R/R/diagnostics.R#L184

The workaround in the meantime is to make sure changepoints are on dates where there are data.

**Comment by bletham** [comment]
OK this was fixed in a519043.

**Comment by bletham** [comment]
Fix is pushed to pypi in v1.0 release
